src/distributions/uniform.rs

```diff
@@ -259,20 +263,61 @@ pub trait UniformSampler: Sized {
         let uniform: Self = UniformSampler::new(low, high);
         uniform.sample(rng)
     }
+
+    /// Sample a single value uniformly from a range with inclusive lower bound
+    /// and inclusive upper bound `[low, high]`.
+    ///
+    /// By default this is implemented using
+    /// `UniformSampler::new_inclusive(low, high).sample(rng)`. However, for
+    /// some types more optimal implementations for single usage may be provided
+    /// via this method.
+    /// Results may not be identical.
+    fn sample_single_inclusive<R: Rng + ?Sized, B1, B2>(low: B1, high: B2, rng: &mut R)
+        -> Self::X
+        where B1: SampleBorrow<Self::X> + Sized,
+              B2: SampleBorrow<Self::X> + Sized
+    {
+        let uniform: Self = UniformSampler::new_inclusive(low, high);
+        uniform.sample(rng)
+    }
 }
 
-impl<X: SampleUniform> From<::core::ops::Range<X>> for Uniform<X> {
+impl<X: SampleUniform> From<Range<X>> for Uniform<X> {
     fn from(r: ::core::ops::Range<X>) -> Uniform<X> {
         Uniform::new(r.start, r.end)
     }
 }
 
-impl<X: SampleUniform> From<::core::ops::RangeInclusive<X>> for Uniform<X> {
+impl<X: SampleUniform> From<RangeInclusive<X>> for Uniform<X> {
     fn from(r: ::core::ops::RangeInclusive<X>) -> Uniform<X> {
         Uniform::new_inclusive(r.start(), r.end())
     }
 }
 
+impl<T: SampleUniform> Distribution<T> for Range<T> {
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> T {
+        T::Sampler::sample_single(&self.start, &self.end, rng)
+    }
+}
+
+impl<T: SampleUniform> Distribution<T> for RangeInclusive<T> {
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> T {
+        T::Sampler::sample_single_inclusive(self.start(), self.end(), rng)
+    }
+}
+
+impl Distribution<usize> for RangeTo<usize> {
```

Comment:
If we are to support this, we should do so for all `T: SampleUniform + Zero` or some such. Unfortunately I think the only way to do this is by adding our own trait `Zero` since we don't really want to use `num-traits`. This should also help clarify that is always means "from zero" (not from the lowest representable value).

Alternatively we could use `T: SampleUniform + From<u8>` which misses support for `i8`, but that's not such an issue.

Reply:
I don't think it's at all correct for this to be implemented with a start point for signed types. I can implement it for all unsigned integer types by exhaustion, though.